### PR TITLE
159972782 servicebus send receive message

### PR DESCRIPTION
**Pull request by Hoagithor**

### CODE REVIEW

File: src/Housing.Selection.Context/ServiceBus/ServiceUserMessaging.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        /// Null address will clear an address.
+        /// An address must be a complete, valid address, or else a 400 error will return.
+        /// Null location will not update.
+        /// If an empty string is passed into location, a 400 error will be returned.
+        /// Location can be 1-255 characters.
+        /// All other fields are ignored.
+        /// </param>
+        public async Task UpdateUserAsync(ApiUser user)
+        {
+            return;
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceUserMessaging.cs]: Code Smell Code Smell: Remove this redundant jump. ([csharpsquid:S3626](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3626))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVjP-rM4Bfu85TPQ1a3&open=AWVjP-rM4Bfu85TPQ1a3)

[ISSUE_KEY]: AWVjP-rM4Bfu85TPQ1a3

File: src/Housing.Selection.Context/ServiceBus/ServiceRoomMessaging.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <param name="room">This is the room to be updated.
+        /// The room must have the ID of the room to be updated,
+        /// but the object only requires fields that need to be
+        /// updated.  Any field that will be left unmodified
+        /// can be left null.
+        /// </param>
+        public async Task UpdateRoomAsync(ApiRoom room)
+        {
+            return;
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceRoomMessaging.cs]: Code Smell Code Smell: Remove this redundant jump. ([csharpsquid:S3626](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3626))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVjP-rI4Bfu85TPQ1a2&open=AWVjP-rI4Bfu85TPQ1a2)

[ISSUE_KEY]: AWVjP-rI4Bfu85TPQ1a2

File: src/Housing.Selection.Context/ServiceBus/ServiceUserMessaging.cs (hunk 1 above)
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceUserMessaging.cs]: Code Smell Code Smell: Remove this redundant jump. ([csharpsquid:S3626](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3626))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVjP-rM4Bfu85TPQ1a3&open=AWVjP-rM4Bfu85TPQ1a3)

[ISSUE_KEY]: AWVjP-rM4Bfu85TPQ1a3

File: src/Housing.Selection.Context/ServiceBus/ServiceRoomMessaging.cs (hunk 2 above)
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceRoomMessaging.cs]: Code Smell Code Smell: Remove this redundant jump. ([csharpsquid:S3626](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3626))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVjP-rI4Bfu85TPQ1a2&open=AWVjP-rI4Bfu85TPQ1a2)

[ISSUE_KEY]: AWVjP-rI4Bfu85TPQ1a2

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        protected async Task<Message> ReceiveMessage()
+        {
+            await Task.Delay(1);
+            Message msg = new Message();
+            //implementation for receiving message from ReceiveQ queue will go here.
+            ReceiveQ.RegisterMessageHandler(
+              async (message, token) =>
+              {
+                  // var userobj = JsonConvert.DeserializeObject(Encoding.UTF8.GetString(message.Body));
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Remove this commented out code. ([csharpsquid:S125](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS125))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVnb_dYXKENQn1ShC1D&open=AWVnb_dYXKENQn1ShC1D)

[ISSUE_KEY]: AWVnb_dYXKENQn1ShC1D

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                },
+                async(exceptionevent) =>
+                {
+                    Debug.WriteLine("Something went wrong");
+                    throw exceptionevent.Exception;
+                });
+
+            DateTime end = DateTime.Now + new TimeSpan(0,0,30);//wait up to 30 seconds
+            while(msg == null && DateTime.Now<end)
+            {
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Either remove or fill this block of code. ([csharpsquid:S108](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS108))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVscJATX2sY87p0cggY&open=AWVscJATX2sY87p0cggY)

[ISSUE_KEY]: AWVscJATX2sY87p0cggY

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 5]
```diff
@@ -26,11 +29,33 @@ public abstract class ServiceMessaging
...
             _queueFactory = queueFactory;
             SendQ = _queueFactory.GetQueueClient(sendQ);
             ReceiveQ = _queueFactory.GetQueueClient(receiveQ);
+            nextId = null;
+            nextMessage = null;
+            ReceiveQ.RegisterMessageHandler(
+                 async (message, token) => {
+                     if (nextId != null && nextMessage == null)
+                     {
+                         if (message.CorrelationId == nextId)
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Merge this if statement with the enclosing one. ([csharpsquid:S1066](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWh&open=AWVsmkIRXKENQn1SqVWh)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWh

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 6]
```diff
@@ -42,14 +67,24 @@ public static Guid GenerateGuid()
...
+        /// </summary>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        protected async Task<Message> SendMessage(Message message)
         {
-            //implementation for receiving message from ReceiveQ queue will go here.
-            return null;
+            await SendQ.SendAsync(message);
+            DateTime end = DateTime.Now + new TimeSpan(0, 1, 0); //wait up to 1 minute
+            while (nextMessage == null && DateTime.Now < end) { } //for the message to get here
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Either remove or fill this block of code. ([csharpsquid:S108](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS108))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWi&open=AWVsmkIRXKENQn1SqVWi)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWi

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs (hunk 5 above)
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Merge this if statement with the enclosing one. ([csharpsquid:S1066](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWh&open=AWVsmkIRXKENQn1SqVWh)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWh

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 7]
```diff
@@ -42,14 +67,25 @@ public static Guid GenerateGuid()
...
+        /// <param name="message"></param>
+        /// <returns></returns>
+        protected async Task<Message> SendMessage(Message message)
         {
-            //implementation for receiving message from ReceiveQ queue will go here.
-            return null;
+            nextId = message.MessageId;
+            await SendQ.SendAsync(message);
+            DateTime end = DateTime.Now + new TimeSpan(0, 1, 0); //wait up to 1 minute
+            while (nextMessage == null && DateTime.Now < end) { } //for the message to get here
```
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Either remove or fill this block of code. ([csharpsquid:S108](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS108))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWi&open=AWVsmkIRXKENQn1SqVWi)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWi

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs (hunk 5 above)
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Merge this if statement with the enclosing one. ([csharpsquid:S1066](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWh&open=AWVsmkIRXKENQn1SqVWh)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWh

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs (hunk 7 above)
Comment (fredbelotte) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Code Smell Code Smell: Either remove or fill this block of code. ([csharpsquid:S108](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS108))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=housing%3Aselection&pullRequest=34&issues=AWVsmkIRXKENQn1SqVWi&open=AWVsmkIRXKENQn1SqVWi)

[ISSUE_KEY]: AWVsmkIRXKENQn1SqVWi

**Comment by Hoagithor** [comment]
Existing TODOs correspond to other stories

File: src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs [hunk 8]
```diff
@@ -42,14 +65,28 @@ public static Guid GenerateGuid()
...
-        {
-            //implementation for sending to SendQ queue will go here.
-        }
-        protected async Task<Message> ReceiveMessage()
+        /// <summary>
+        /// Send a message and return its response. Should never return null. Exception thrown in timout, time set here.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        protected async Task<Message> SendMessage(Message message)
```
Comment (ajzoell) [inline src/Housing.Selection.Context/ServiceBus/ServiceMessaging.cs]: Can you fill in the param and returns tags? Param is just to explain what the parameter you pass in is. Returns explains what you're returning.

**Pull request merged**